### PR TITLE
feat(ecmascript): implement array constructor builtins

### LINKED ISSUE #143: Array constuctor builtin functions

### PR DESCRIPTION
**Pull request by eliassjogreen**
This PR implements the array constructor builtin functions according to spec.

Closes #143

### CODE REVIEW

File: nova_vm/src/ecmascript/builtins/indexed_collections/array_objects/array_constructor.rs [hunk 1]
```diff
@@ -60,47 +82,293 @@ impl BuiltinGetter for ArrayGetSpecies {
...
+            array_create(agent, len, len, None)?.into_object()
+        };
+
+        // 11. Let k be 0.
+        let mut k = 0;
+
+        // 12. Repeat, while k < len,
+        while k < len {
+            // 𝔽(k)
+            let fk = Value::try_from(k as f64).unwrap();
```
Comment (aapoalas) [inline nova_vm/src/ecmascript/builtins/indexed_collections/array_objects/array_constructor.rs:334]: **issue:** Same stuff here as above, let's keep using integers and not do the string conversion for property keys.

File: nova_vm/src/ecmascript/builtins/indexed_collections/array_objects/array_constructor.rs [hunk 2]
```diff
@@ -109,16 +377,66 @@ impl ArrayConstructor {
...
 
-    fn of(_agent: &mut Agent, _this_value: Value, _arguments: ArgumentsList) -> JsResult<Value> {
-        todo!();
+    /// ### [23.1.2.3 Array.of ( ...items )](https://tc39.es/ecma262/#sec-array.of)
+    fn of(agent: &mut Agent, this_value: Value, arguments: ArgumentsList) -> JsResult<Value> {
+        // 1. Let len be the number of elements in items.
+        let len = arguments.len();
+
+        // 2. Let lenNumber be 𝔽(len).
+        let len_number = Value::try_from(len as f64).unwrap();
```
Comment (aapoalas) [inline nova_vm/src/ecmascript/builtins/indexed_collections/array_objects/array_constructor.rs:387]: **issue:** Same as above, keep integer.

File: nova_vm/src/ecmascript/builtins/indexed_collections/array_objects/array_constructor.rs [hunk 3]
```diff
@@ -109,16 +377,66 @@ impl ArrayConstructor {
...
+
+        // 3. Let C be the this value.
+        // 4. If IsConstructor(C) is true, then
+        let a = if is_constructor(agent, this_value) {
+            // SAFETY: Only Functions can be constructors
+            let constructor = unsafe { Function::try_from(this_value).unwrap_unchecked() };
```
Comment (aapoalas) [inline nova_vm/src/ecmascript/builtins/indexed_collections/array_objects/array_constructor.rs:393]: **issue:** Same as above, unwrap normally.

File: nova_vm/src/ecmascript/builtins/indexed_collections/array_objects/array_constructor.rs [hunk 4]
```diff
@@ -109,16 +377,66 @@ impl ArrayConstructor {
...
+        let mut k = 0;
+
+        // 7. Repeat, while k < len,
+        while k < len {
+            // a. Let kValue be items[k].
+            let k_value = arguments.get(k);
+
+            // b. Let Pk be ! ToString(𝔽(k)).
+            let pk =
+                PropertyKey::from(to_string(agent, Value::try_from(k as f64).unwrap()).unwrap());
```
Comment (aapoalas) [inline nova_vm/src/ecmascript/builtins/indexed_collections/array_objects/array_constructor.rs]: **issue:** Same as above, keep integer and avoid string conversion.

**Review by aapoalas — CHANGES_REQUESTED** [review]
Awesome stuff, thank you for doing this! <3

There's unfortunately a string/integer PropertyKey thing here that we should resolve, caused by a bug in the engine. I left `issue:` comments in those places that should be changed before merging. Note: The comment "notation" comes from [Conventional Comments](https://conventionalcomments.org/).

Anyway, absolutely awesome stuff, thank you so much! <3

File: nova_vm/src/ecmascript/abstract_operations/operations_on_iterator_objects.rs [hunk 5]
```diff
@@ -230,7 +262,7 @@ pub(crate) fn iterator_close(
     todo!()
 }
 
-/// ### [7.4.9 IfAbruptCloseIterator ( value, iteratorRecord )](https://tc39.es/ecma262/#sec-ifabruptcloseiterator)
+/// ### [7.4.10 IfAbruptCloseIterator ( value, iteratorRecord )](https://tc39.es/ecma262/#sec-ifabruptcloseiterator)
```
[the remark this answers is not in the document]
Comment (eliassjogreen) [reply nova_vm/src/ecmascript/abstract_operations/operations_on_iterator_objects.rs:265]: It's `7.4.8 IteratorStepValue` that is new I believe :)

File: nova_vm/src/ecmascript/abstract_operations/type_conversion.rs [hunk 6]
```diff
@@ -618,7 +618,7 @@ pub(crate) fn canonical_numeric_index_string(
     argument: String,
 ) -> Option<Number> {
     // 1. If argument is "-0", return -0𝔽.
-    if argument == BUILTIN_STRING_MEMORY._0 {
+    if argument == BUILTIN_STRING_MEMORY.__0 {
```
Comment (aapoalas) [inline nova_vm/src/ecmascript/abstract_operations/type_conversion.rs:621]: **issue:** This seems wrong: The string `-0` should get turned into `_0`; `__0` would be eg. `--0`.

**Review by aapoalas — APPROVED** [review]
Ooops I was wrong about the string logic. Looks good to me!

Merging after CI says a-okay. Thank you so much <3

**Pull request merged**